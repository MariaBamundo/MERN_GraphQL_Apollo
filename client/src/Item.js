import React, {Component} from 'react';
import gql from 'graphql-tag';
import { graphql, Mutation, compose } from 'react-apollo';
import { storeKeyNameFromField } from 'apollo-utilities';


class Item extends Component
{
    constructor(props)
    {
        super(props);
        this.state = {
            tasks: this.props.tasks
        };
        this.deleteTask.bind(this.props.deleteTask);
        this.checkTask.bind(this.props.checkTask);
        this.changeText.bind(this.props.changeText);
    }


    checkTask(task)
    {
        this.props.checkTask(task);
    }

    editButton(task)
    {
        var curElement = document.getElementById(task.id);
        var newText = curElement.innerText;
        var _this = this;
        curElement.contentEditable = "true";
        curElement.focus();
        curElement.addEventListener('keypress', function(e)
        {
            var key = e.which || e.keyCode;
            if(key === 13)
            {
                e.target.contentEditable = "false";
                e.target.blur();
                newText = e.target.innerText;
                _this.changeText(task, newText);
            }
        });
        curElement.addEventListener('blur', function(e)
        {
            e.target.contentEditable = "false";
            newText = e.target.innerText;
            _this.changeText(task,newText);
            
        });
    }
    changeText(task,newText)
    {
        this.props.changeText(task,newText)
    }

    deleteTask(task)
    {
        this.props.deleteTask(task);
    }

    render()
    {
        //console.log("items", this.state.tasks);
        return(
            this.props.tasks.map((currentTask)=>
                <tr key={currentTask.id}>
                    <td>
                        <input type="checkbox" checked={currentTask.isDone} className = "checkbox" onChange={() => this.checkTask(currentTask)} />
                        <p id={currentTask.id}>{currentTask.item}</p>
                        <button onClick={()=> this.deleteTask(currentTask)}> Delete Item </button>
                        <button onClick={()=> this.editButton(currentTask)}> Edit Item</button>
                    </td>
                </tr>
            )
        );
    }
}


export default Item;