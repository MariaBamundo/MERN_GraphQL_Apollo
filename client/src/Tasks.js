import React, {Component} from 'react';
import { Query } from "react-apollo";
import gql from 'graphql-tag';
import Item from './Item';
import {graphql, compose} from 'react-apollo';

class Tasks extends Component
{
    constructor(props)
    {
        super(props);
        this.state = {
            term: '',
            error: null,
            tasks: this.props.data
        };
        this.handleSubmit = this.handleSubmit.bind(this);
        this.changeText = this.changeText.bind(this);
        //console.log(this.state.tasks);
        
    }

    handleSubmit(e) 
    {
        e.preventDefault();
        if(this.state.term.replace(/^\s+/g, '').length < 1)
        {
            var warning = document.getElementById('warning');
            warning.style.visibility = "visible";
            this.setState(
            {
                term: ''
            });
            return;
        }
        else
        {
            var warning = document.getElementById('warning');
            warning.style.visibility = "hidden";

            this.props.addItem(
            {
                variables:
                {
                    item: this.state.term,
                    isDone: false
                },
                update: (store, {data: {addItem}}) => 
                {
                    this.setState({
                        term: '',
                        tasks: [...this.state.tasks,addItem]
                    });
                }
            }).then(function getResponse(response){
                console.log(response);
            });
        }
    }
    onChange = (event) => 
    {
        this.setState({term: event.target.value});
    }

    changeText(task,newText)
    {
        this.props.updateItem({
            variables:{
                id: task.id,
                item: newText,
                isDone: task.isDone
            },
            update: (store, {data:{updateItem}}) =>
            {
                const data = this.state.tasks.slice();
                const index = data.findIndex(task => task.id === updateItem.id);
                data[index] = updateItem;
                this.setState({
                    tasks: data
                });
            }
        });
    }

    checkTask(task)
    {
        this.props.updateItem({
            variables:{
                id: task.id,
                item: task.item,
                isDone: !task.isDone
            },
            update: (store, {data:{updateItem}}) =>
            {
                const data = this.state.tasks.slice();
                const index = data.findIndex(task => task.id === updateItem.id);
                data[index] = updateItem;
                this.setState({
                    tasks: data
                });
            }
        });
    }

    deleteTask(task)
    {
        this.props.removeItem(
        {
            variables:
            {
                id: task.id
            },
        }).then(function getResponse(response)
        {
            console.log(response);
        });
        this.setState(prevState => ({
            tasks: prevState.tasks.filter(el => el != task)
        }));
        //console.log(this.state.tasks);
        
    }

    render()
    {
        //console.log("GOT HERE", this.state.tasks);
        return(
            <div className="dataCenter">
                <form className="table" onSubmit={this.handleSubmit}>
                    <input value={this.state.term} onChange={this.onChange}/>
                    <p id="warning"> Please type in a unique item to add it to the To-Do list!</p>
                    <button>Submit</button>
                </form>
                <div className="tableDiv">
                    <table className="table-bordered table-hover">
                        <tbody>
                            <Item tasks={this.state.tasks} deleteTask={this.deleteTask.bind(this)} checkTask={this.checkTask.bind(this)} changeText={this.changeText.bind(this)}/>
                        </tbody>
                    </table>
                </div>
            </div>
        );
    }
}

const taskQuery = gql`
  query todoList {
    tasks {
      id
      item
      isDone
    }
  }`;

const deleteItem = gql`
    mutation removeItem($id: String!){
        removeItem(id:$id){
            id
            item
            isDone
        }
    }`;

const ADD_TODO = gql`
    mutation addItem($item: String!, $isDone: Boolean!){
      addItem(item: $item, isDone: $isDone){
        id
        item
        isDone
      }
    }`;

const markCompletedQuery = gql`
    mutation updateItem($id: String!, $item: String! $isDone: Boolean!) {
        updateItem(id: $id, item: $item, isDone: $isDone){
            id
            item
            isDone
        }
    }`;


export default compose(graphql(deleteItem,{name: 'removeItem'}),
graphql(taskQuery,{name: 'getTasks'}),
graphql(ADD_TODO,{name: 'addItem'}),graphql(markCompletedQuery,{name: 'updateItem'}))(Tasks);
