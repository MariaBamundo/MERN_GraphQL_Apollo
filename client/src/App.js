import React, { Component } from 'react';
import logo from './logo.svg';
import './App.css';
import ApolloClient from 'apollo-boost';
import { ApolloProvider, Mutation } from 'react-apollo';
import Tasks from './Tasks';
import gql from 'graphql-tag';

const client = new ApolloClient({
  uri: "http://localhost:4000/graphql",
});

class App extends Component 
{
  constructor(props)
  {
    super(props);
    this.state = {
      term: '',
      error: null,
      data: []
    };
  }


  onSubmit = (event) => 
  {
    event.preventDefault();
    if(this.state.term.replace(/^\s+/g, '').length < 1)
    {
      var warning = document.getElementById('warning');
      warning.style.visibility = "visible";
      this.setState(
      {
        term: ''
      });
            
    }
    else
    {
      //THIS IS WHERE YOU WERE LAST TRYING TO GET THE ADDITEM MUTATION WORKING 
      var warning = document.getElementById('warning');
      warning.style.visibility = "hidden";
      <Mutation mutation={ADD_TODO}>
      {(addItem, {data})=> (
        addItem({variables:{item:this.state.term,isDone:false}})
      )}
      </Mutation>
      this.setState(
      {
        term: ''
      });
    }
  }



  onChange = (event) => 
  {
    this.setState({term: event.target.value});
  }



  render() 
  {
    return(
      <ApolloProvider client={client}>
        <div className="App">
          <h1>To-Do List: </h1>
          <form className="table" onSubmit={this.onSubmit}>
            <input value={this.state.term} onChange={this.onChange}/>
            <p id="warning"> Please type in a unique item to add it to the To-Do list!</p>
            <button>Submit</button>
          </form>
          <div className="tableDiv">
            <table className="table-bordered table-hover">
              <tbody>
                <Tasks />
              </tbody>
            </table>
          </div>
        </div>
      </ApolloProvider>
    );
  }
}

const ADD_TODO = gql`
  mutation addItem($item: String!, $isDone: Boolean!){
    addItem(item: $item, isDone: $isDone){
      id
      item
      isDone
    }
  }
`;

export default App;
