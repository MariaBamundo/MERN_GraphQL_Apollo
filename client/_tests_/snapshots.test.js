import React from 'react';
import Tasks from '../src/Tasks'
import renderer from 'react-test-renderer';
import {ApolloProvider} from 'react-apollo';
import {ApolloClient} from 'apollo-client';
import {HttpLink} from 'apollo-link-http';
import {InMemoryCache} from 'apollo-cache-inmemory';
import fetch from 'node-fetch';
import Item from '../src/Item';
import App from '../src/App';

const client = new ApolloClient({
    link: new HttpLink({
        uri: 'http://localhost:4000/graphql',
        fetch,
    }),
    cache: new InMemoryCache()
});

const tasks = [
    {
      "id": "5b5f145e605ec80c819f5824",
      "item": "one",
      "isDone": false
    }
  ];


test('Snapshot of Tasks component', () => {
    const component = renderer.create(
        <ApolloProvider client={client}>
            <Tasks data={tasks}/>
        </ApolloProvider>
    );
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
    //console.log("TASKS", tree);
    const form = tree.children[0];
    const itemDiv = tree.children[1];
    //console.log(itemDiv.children[0].children[0].children[0].children[0].children[0]);
    const itemCheckbox = itemDiv.children[0].children[0].children[0].children[0].children[0];

    //console.log(deleteButton)
    
    itemCheckbox.props.onChange();
    tree = component.toJSON();
    expect(tree).toMatchSnapshot();

    const deleteButton = itemDiv.children[0].children[0].children[0].children[0].children[2];
    deleteButton.props.onClick();
    tree = component.toJSON();
    expect(tree).toMatchSnapshot();

    console.log(form);
    //e = null;
    const textBox = form.children[0];
    console.log(textBox);
    //textBox.props.onChange();
    //form.props.onSubmit();
    tree = component.toJSON();
    expect(tree).toMatchSnapshot();
});

test('Snapshot of App component', () => {
    const component = renderer.create(
        <ApolloProvider client={client}>
            <App />
        </ApolloProvider>
    );
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
    //console.log(tree);
    //const itemChild = tree.children[0].children[0].children[0];
    //console.log("ITEMCHILD ",itemChild);
});

test('Snapshot of Item component', () => {
    const component = renderer.create(
        <ApolloProvider client={client}>
            <Item tasks={tasks}/>
        </ApolloProvider>
    );
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
    //console.log(tree);

    //console.log(tree.children[0].children);
    const child = tree.children[0].children[0];
    //console.log("CHILD: ",child);
    //child.props.onChange();
    tree = component.toJSON();
    expect(tree).toMatchSnapshot();
})