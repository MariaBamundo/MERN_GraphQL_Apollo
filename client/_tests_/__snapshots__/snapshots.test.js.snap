// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snapshot of App component 1`] = `
<div>
   loading 
</div>
`;

exports[`Snapshot of Item component 1`] = `
<tr>
  <td>
    <input
      checked={false}
      className="checkbox"
      onChange={[Function]}
      type="checkbox"
    />
    <p
      id="5b5f145e605ec80c819f5824"
    >
      one
    </p>
    <button
      className="delete"
      onClick={[Function]}
    >
       Delete Item 
    </button>
    <button
      className="edit"
      onClick={[Function]}
    >
       Edit Item
    </button>
  </td>
</tr>
`;

exports[`Snapshot of Item component 2`] = `
<tr>
  <td>
    <input
      checked={false}
      className="checkbox"
      onChange={[Function]}
      type="checkbox"
    />
    <p
      id="5b5f145e605ec80c819f5824"
    >
      one
    </p>
    <button
      className="delete"
      onClick={[Function]}
    >
       Delete Item 
    </button>
    <button
      className="edit"
      onClick={[Function]}
    >
       Edit Item
    </button>
  </td>
</tr>
`;

exports[`Snapshot of Tasks component 1`] = `
<div
  className="dataCenter"
>
  <form
    className="table"
    onSubmit={[Function]}
  >
    <input
      onChange={[Function]}
      value=""
    />
    <p
      id="warning"
    >
       Please type in a unique item to add it to the To-Do list!
    </p>
    <button>
      Submit
    </button>
  </form>
  <div
    className="tableDiv"
  >
    <table
      className="table-bordered table-hover"
    >
      <tbody>
        <tr>
          <td>
            <input
              checked={false}
              className="checkbox"
              onChange={[Function]}
              type="checkbox"
            />
            <p
              id="5b5f145e605ec80c819f5824"
            >
              one
            </p>
            <button
              className="delete"
              onClick={[Function]}
            >
               Delete Item 
            </button>
            <button
              className="edit"
              onClick={[Function]}
            >
               Edit Item
            </button>
          </td>
        </tr>
      </tbody>
    </table>
  </div>
</div>
`;

exports[`Snapshot of Tasks component 2`] = `
<div
  className="dataCenter"
>
  <form
    className="table"
    onSubmit={[Function]}
  >
    <input
      onChange={[Function]}
      value=""
    />
    <p
      id="warning"
    >
       Please type in a unique item to add it to the To-Do list!
    </p>
    <button>
      Submit
    </button>
  </form>
  <div
    className="tableDiv"
  >
    <table
      className="table-bordered table-hover"
    >
      <tbody>
        <tr>
          <td>
            <input
              checked={false}
              className="checkbox"
              onChange={[Function]}
              type="checkbox"
            />
            <p
              id="5b5f145e605ec80c819f5824"
            >
              one
            </p>
            <button
              className="delete"
              onClick={[Function]}
            >
               Delete Item 
            </button>
            <button
              className="edit"
              onClick={[Function]}
            >
               Edit Item
            </button>
          </td>
        </tr>
      </tbody>
    </table>
  </div>
</div>
`;

exports[`Snapshot of Tasks component 3`] = `
<div
  className="dataCenter"
>
  <form
    className="table"
    onSubmit={[Function]}
  >
    <input
      onChange={[Function]}
      value=""
    />
    <p
      id="warning"
    >
       Please type in a unique item to add it to the To-Do list!
    </p>
    <button>
      Submit
    </button>
  </form>
  <div
    className="tableDiv"
  >
    <table
      className="table-bordered table-hover"
    >
      <tbody />
    </table>
  </div>
</div>
`;

exports[`Snapshot of Tasks component 4`] = `
<div
  className="dataCenter"
>
  <form
    className="table"
    onSubmit={[Function]}
  >
    <input
      onChange={[Function]}
      value=""
    />
    <p
      id="warning"
    >
       Please type in a unique item to add it to the To-Do list!
    </p>
    <button>
      Submit
    </button>
  </form>
  <div
    className="tableDiv"
  >
    <table
      className="table-bordered table-hover"
    >
      <tbody />
    </table>
  </div>
</div>
`;
